tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -536,6 +531,356 @@ def estimate_affine_transform(
+
+        This function removes the duplicated points and the points
+        which are identified outside the tissue region.
+
+        Args:
+            fixed_mask (:class:`numpy.ndarray`):
+                A binary tissue mask for the fixed image.
+            moving_mask (:class:`numpy.ndarray`):
+                A binary tissue mask for the moving image.
+            fixed_matched_points (:class:`numpy.ndarray`):
+                (N, 2) array of coordinates.
+            moving_matched_points (:class:`numpy.ndarray`):
+                (N, 2) array of coordinates.
+            quality (:class:`numpy.ndarray`):
+                An array representing quality of each matching point.
+
+        Returns:
+            tuple:
+                - np.ndarray - Filtered matching points for a fixed image.
+                - np.ndarray - Filtered matching points for a moving image.
+                - np.ndarray - Quality of matching points.
+
+        """
+        included_indx = self.find_points_inside_boundary(
+            fixed_mask, fixed_matched_points
+        )
+        fixed_matched_points, moving_matched_points, quality = (
+            fixed_matched_points[included_indx, :],
+            moving_matched_points[included_indx, :],
+            quality[included_indx],
+        )
+        included_indx = self.find_points_inside_boundary(
+            moving_mask, moving_matched_points
+        )
+        fixed_matched_points, moving_matched_points, quality = (
+            fixed_matched_points[included_indx, :],
+            moving_matched_points[included_indx, :],
+            quality[included_indx],
+        )
+
+        # remove duplicate matching points
+        duplicate_ind = []
+        unq, count = np.unique(fixed_matched_points, axis=0, return_counts=True)
+        repeated_points = unq[count > 1]
+        for repeated_point in repeated_points:
+            repeated_idx = np.argwhere(
+                np.all(fixed_matched_points == repeated_point, axis=1)
+            )
+            duplicate_ind = np.hstack([duplicate_ind, repeated_idx.ravel()])
+
+        unq, count = np.unique(moving_matched_points, axis=0, return_counts=True)
+        repeated_points = unq[count > 1]
+        for repeated_point in repeated_points:
+            repeated_idx = np.argwhere(
+                np.all(moving_matched_points == repeated_point, axis=1)
+            )
+            duplicate_ind = np.hstack([duplicate_ind, repeated_idx.ravel()])
+
+        if len(duplicate_ind) > 0:
+            duplicate_ind = duplicate_ind.astype(int)
+            fixed_matched_points = np.delete(
+                fixed_matched_points, duplicate_ind, axis=0
+            )
+            moving_matched_points = np.delete(
+                moving_matched_points, duplicate_ind, axis=0
+            )
+            quality = np.delete(quality, duplicate_ind)
+
+        return fixed_matched_points, moving_matched_points, quality
+
+    def register(
```

Comment:
This function could do with some comments explaining what is going on. Also, a bit of a general tidy up. Maybe better variable names, or you may be able to extract some parts into another function to make this one less complex.